accord-core/src/main/java/accord/coordinate/tracking/AbstractResponseTracker.java

```diff
@@ -0,0 +1,102 @@
+package accord.coordinate.tracking;
+
+import accord.local.Node;
+import accord.topology.Shard;
+import accord.topology.Shards;
+import com.google.common.annotations.VisibleForTesting;
+
+import java.util.*;
+import java.util.function.BiConsumer;
+import java.util.function.BiFunction;
+import java.util.function.Predicate;
+
+abstract class AbstractResponseTracker<T extends AbstractResponseTracker.ShardTracker>
+{
+    private final Shards shards;
+    private final Map<Node.Id, List<T>> nodeMap = new HashMap<>();
+    private final T[] trackers;
+
+    static class ShardTracker
+    {
+        final Shard shard;
+
+        public ShardTracker(Shard shard)
+        {
+            this.shard = shard;
+        }
+    }
+
+    private static <T extends ShardTracker> void indexNodes(T tracker, Map<Node.Id, List<T>> nodeMap, int maxTrackers)
+    {
+        List<Node.Id> nodes = tracker.shard.nodes;
+        for (int i=0, mi=nodes.size(); i<mi; i++)
+        {
+            Node.Id node = nodes.get(i);
+            List<T> trackers = nodeMap.get(node);
+            if (trackers == null)
+            {
+                trackers = new ArrayList<>(maxTrackers);
+                nodeMap.put(node, trackers);
+            }
+            trackers.add(tracker);
+        }
+    }
+
+    public AbstractResponseTracker(Shards shards)
+    {
+        this.shards = shards;
+        trackers = createInfoArray(shards.size());
```

Comment:
I wanted to avoid taking up space on the heap for something we can define at compile time. I replaced the call with `(T[]) new Object[shards.size()];` instead

Reply:
I don't think this should take up any space? It should be a static function so instantiated only once. It's likely it will be inlined inside each of the child constructors, even.

Reply:
AFAICT this is only used in the constructor right now? Perhaps you have some follow on work that re-allocates the array?

Anyway, this isn't super important and will likely be refactored again in future. So don't worry about it.